### PR TITLE
Change the "sticky" attribute to take a direction

### PR DESCRIPTION
**Pull request by efeyakinci**
This change closes https://github.com/typst/typst/issues/5259

It is useful to be able to create blocks that stick to the element above them. The most notable use case is for lists and enums to prevent pagebreaks in documents like
```typ
//...
The following list shows the funding sources that sponsored this article:
// A pagebreak is undesirable here!
- Acme Inc.
- Typst GmbH
- Company C
```

Other use cases include code blocks that follow text, or footnotes such as table footnotes that refer to content above them

Currently, there is no good way to achieve this. This change modifies the sticky parameter to take one of `"below", "above", "both", or none.`

This change also includes a change to the behavior of sticky when interacting with breakable blocks that spill that becomes more visible with such use cases. For an example like the following:

```typ
#set page(height:120pt, margin: (y: 0pt))
#set block(spacing: 0pt, width: 20pt)

#block(height: 60pt, fill: red)[A]
#block(height: 70pt, fill: green, breakable: true, sticky: true)[A]
#block(height: 10pt, fill: purple)[C]
```

The behavior is that the middle block will cause a pagebreak so that it can be with the following block in unbroken form. However, this is likely not the least-surprising behavior. For example, consider a table that covers 90% of the height of a page. If a user added such a table where 20% of the vertical space is occupied and made the table sticky to add footnotes after the table, the expected behavior is likely not that the first page is left 90% blank, and the second page contains the table and footnotes in their entirety.

This PR adds a check to cause a pagebreak on a sticky elements that ends the page. In this case, we break if and only if there is no spill. Examining the cases, we see

1.  There is no spill: Then, we should break, so that the element can stay with the element following it, unless the element is the first element on the page.
1. There is some spill. Clearly (given that one agrees with the general behavior), we not should break if the spillage can be next to the element on the next page. However, this change does make it possible where we avoid breaking, even though breaking would cause spillage onto a third page whose size can accommodate the successor after it.

The second (rare) edge case does allow for a sticky element to be on a different page than its successor, although the behavior with the change is still (I believe) less surprising than the current behavior.

This is non-trivial to fix, as knowing when to break would require calculating the amount of spillage on the final page of spillage, knowing the size of that page, and calculating whether the items will fit. As such, this is not done in this PR.

### CODE REVIEW

File: crates/typst-layout/src/flow/distribute.rs [hunk 1]
```diff
@@ -443,7 +472,13 @@ impl<'a, 'b> Distributor<'a, 'b, '_, '_, '_> {
...
             // suffix to the next region.
             if let Some(snapshot) = self.sticky.take() {
-                self.restore(snapshot)
+                // Only restore snapshot if there's no spill
+                // If a sticky breakable element can be spilled to the following page,
+                // it's fine to put some of it on this page, since the next page will still
+                // have the next element on the same page as *some* of the spilled content.
+                if self.composer.work.spill.is_none() {
+                    self.restore(snapshot)
+                }
```
Comment (PgBiel) [inline crates/typst-layout/src/flow/distribute.rs:481]: Can you add a test where this changes the behavior of existing `sticky: "below"` blocks so we can see what this is actually changing? Want to avoid regressions here.

File: crates/typst-library/src/layout/container.rs [hunk 2]
```diff
@@ -338,23 +359,29 @@ pub struct BlockElem {
...
     /// #lorem(10)
+    ///
+    /// #table(
+    /// columns: 2,
+    /// [A], [B],
+    /// [C], [D],
+    /// )
+    /// #block(sticky: "above")[The above table shows that...]
     /// ```
-    #[default(false)]
```
Comment (PgBiel) [inline crates/typst-library/src/layout/container.rs:356]: Let's add some compatibility behavior to temporarily accept bools with a warning. (You can leave this for last, after all other reviews have been addressed.)

File: crates/typst-layout/src/flow/distribute.rs [hunk 3]
```diff
@@ -324,10 +325,28 @@ impl<'a, 'b> Distributor<'a, 'b, '_, '_, '_> {
...
+                    single.sticky.as_ref().map(|s| s.is_sticky_above()).unwrap_or(false),
+                ),
+                Child::Multi(multi) => Some(
+                    multi.sticky.as_ref().map(|s| s.is_sticky_above()).unwrap_or(false),
+                ),
+                _ => None,
+            })
+            .unwrap_or(false);
+
+        let mut stick_to_successor = || {
```
Comment (PgBiel) [inline crates/typst-layout/src/flow/distribute.rs:349]: I'd love some tests showing how a sequence of sticky: "above" which don't fit in a single page behave. There is one already for sticky: "below", not too sure if it was `block-sticky-many`, but might have been

**Review by PgBiel — COMMENTED** [review]
Thanks for the proposal. Regarding the feature itself, I can see there being value in being able to take control over "widow prevention", similar to how the existing sticky attribute allows control over "orphan prevention".

However, I have a few concerns; I'll list some below to start, but in general, one important concern is that a single test won't take care of this change with potential for regressions. I feel like we need to clearly study, document and test the behavior of each possible flow child when affected by `sticky: "above"`, and perhaps add more information about edge cases - I mention some below.

**Comment by laurmaedje** [comment]
@efeyakinci Do you still plan to work on this? In principle, it would not be exceedingly hard to implement this feature, but it would need some changes from the current implementation. As it is very central code, it would also need to be well-tested. If you'd prefer not to proceed, that's totally fine, too. Then, I'd close the PR, to "unlock the mutex" on the issue, so to speak.

Regarding API design, I posted some thoughts in the issue: https://github.com/typst/typst/issues/5259#issuecomment-3242195471

Implementation-wise, my gut feeling is that it would be easier to move the pre-processing of this to the collection phase (`collect.rs`) as it should be possible to make this transformation without knowledge of the concrete layout regions. The `LineChild` would then also need a `sticky` bool that could be set by a following Single/Multi child with `sticky: top`. (Take all of this with a graint of salt, as I've not actually experimented with it.)

**Comment by laurmaedje** [comment]
For now, I'll close this due to inactivity. Thanks still!
